components/brave_shields/browser/ad_block_service_browsertest.cc

```diff
@@ -744,3 +744,112 @@ IN_PROC_BROWSER_TEST_F(AdBlockServiceTest, CancelRequestOptionTest) {
+                                          &as_expected));
+  EXPECT_TRUE(as_expected);
+}
+
+// Test cosmetic filtering on elements added dynamically
+IN_PROC_BROWSER_TEST_F(AdBlockServiceTest, CosmeticFilteringDynamic) {
+  UpdateAdBlockInstanceWithRules(
+      "##.blockme"
+  );
+
+  GURL tab_url = embedded_test_server()->GetURL("b.com",
+                                                "/cosmetic_filtering.html");
+  ui_test_utils::NavigateToURL(browser(), tab_url);
+
+  content::WebContents* contents =
+    browser()->tab_strip_model()->GetActiveWebContents();
+
+  bool as_expected = false;
+  ASSERT_TRUE(ExecuteScriptAndExtractBool(contents,
+                                          "addElementsDynamically();\n"
+                                          "checkSelector('.blockme', 'display', 'none')",
+                                          &as_expected));
+  EXPECT_TRUE(as_expected);
+
+  as_expected = false;
+  ASSERT_TRUE(ExecuteScriptAndExtractBool(contents,
+                                          "checkSelector('.dontblockme', 'display', 'block')",
+                                          &as_expected));
+  EXPECT_TRUE(as_expected);
+}
+
+// Test custom style rules
+IN_PROC_BROWSER_TEST_F(AdBlockServiceTest, CosmeticFilteringCustomStyle) {
+  UpdateAdBlockInstanceWithRules(
+      "b.com##.ad:style(padding-bottom: 0px)"
+  );
+
+  GURL tab_url = embedded_test_server()->GetURL("b.com",
+                                                "/cosmetic_filtering.html");
+  ui_test_utils::NavigateToURL(browser(), tab_url);
+
+  content::WebContents* contents =
+    browser()->tab_strip_model()->GetActiveWebContents();
+
+  bool as_expected = false;
+  ASSERT_TRUE(ExecuteScriptAndExtractBool(contents,
+                                          "checkSelector('.ad', 'padding-bottom', '0px')",
+                                          &as_expected));
+  EXPECT_TRUE(as_expected);
+}
+
+// Test rules overridden by hostname-specific exception rules
+IN_PROC_BROWSER_TEST_F(AdBlockServiceTest, CosmeticFilteringUnhide) {
+  UpdateAdBlockInstanceWithRules(
+      "##.ad\n"
+      "b.com#@#.ad\n"
+      "###ad-banner\n"
+      "a.com#@##ad-banner"
+  );
+
+  GURL tab_url = embedded_test_server()->GetURL("b.com",
+                                                "/cosmetic_filtering.html");
+  ui_test_utils::NavigateToURL(browser(), tab_url);
+
+  content::WebContents* contents =
+    browser()->tab_strip_model()->GetActiveWebContents();
+
+  bool as_expected = false;
+  ASSERT_TRUE(ExecuteScriptAndExtractBool(contents,
+                                          "checkSelector('.ad', 'display', 'block')",
+                                          &as_expected));
+  EXPECT_TRUE(as_expected);
+
+  as_expected = false;
+  ASSERT_TRUE(ExecuteScriptAndExtractBool(contents,
+                                          "checkSelector('#ad-banner', 'display', 'none')",
+                                          &as_expected));
+  EXPECT_TRUE(as_expected);
```

Comment:
Good job on these tests 👍

Reply:
Could you add a test for when cosmetic filters are off that it doesn't actually do any work?
Either by preference or by calling this explicitly:
```
brave_shields::SetCosmeticFilteringControlType(profile_,
                                  value ? ControlType::BLOCK
                                        : ControlType::ALLOW,
                                  GURL());
```
I think setting just the pref won't work currently because this is done in the callback and the pref isn't actually checked anywhere.

Reply:
For the reason I mentioned [above](https://github.com/brave/brave-core/pull/3303#discussion_r348112733), I couldn't actually get the value of the preference when running tests, so I defaulted to having cosmetic filtering on. This would be a good thing to fix and test, but I don't know where to go from here.

Reply:
Since moving to a feature flag I was able to make a test for this.